index.ts

```diff
@@ -0,0 +1,245 @@
+    const releaseLock = await globalMutex.acquire()
+    return { releaseLock }
+  }
+
+  /**
+  this will return an object with the `nextNonce` `nonceDetails`, and the releaseLock
+  Note: releaseLock must be called after adding a signed tx to pending transactions (or discarding).
+
+  @param address the hex string for the address whose nonce we are calculating
+  @returns {Promise<NonceLock>}
+  */
+  async getNonceLock (address: string): Promise<NonceLock> {
+    // await global mutex free
+    await this._globalMutexFree()
+    // await lock free, then take lock
+    const releaseLock = await this._takeMutex(address)
+    try {
+      // evaluate multiple nextNonce strategies
+      const networkNonceResult = await this._getNetworkNextNonce(address)
+      const highestLocallyConfirmed = this._getHighestLocallyConfirmed(address)
+      const nextNetworkNonce = networkNonceResult.nonce
+      const highestSuggested = Math.max(nextNetworkNonce, highestLocallyConfirmed)
+
+      const pendingTxs: Transactions[] = this.getPendingTransactions(address)
+      const localNonceResult = this._getHighestContinuousFrom(pendingTxs, highestSuggested)
+
+      const nonceDetails:NonceDetails = {
+        params:{
+          highestLocallyConfirmed:highestLocallyConfirmed,
+          nextNetworkNonce:nextNetworkNonce,
+          highestSuggested: highestSuggested,
+        },
+        local: localNonceResult,
+        network: networkNonceResult
+      }
+
+      const nextNonce = Math.max(networkNonceResult.nonce, localNonceResult.nonce)
+      assert(Number.isInteger(nextNonce), `nonce-tracker - nextNonce is not an integer - got: (${typeof nextNonce}) "${nextNonce}"`)
+
+      // return nonce and release cb
+      return { nextNonce, nonceDetails, releaseLock }
+    } catch (err) {
+      // release lock if we encounter an error
+      releaseLock()
+      throw err
+    }
+  }
+
+  async _globalMutexFree () {
+    const globalMutex = this._lookupMutex('global')
+    const releaseLock = await globalMutex.acquire()
+    releaseLock()
+  }
+
+  async _takeMutex (lockId: string): Promise<any> {
+    const mutex = this._lookupMutex(lockId)
+    const releaseLock = await mutex.acquire()
+    return releaseLock
+  }
+
+  _lookupMutex (lockId: string): any {
+    let mutex = this.lockMap[lockId]
+    if (!mutex) {
+      mutex = new Mutex()
+      this.lockMap[lockId] = mutex
+    }
+    return mutex
+  }
+
+  /**
+   * Function returns the nonce details from teh network based on the latest block 
+   * and eth_getTransactionCount method
+   * 
+   * @param address the hex string for the address whose nonce we are calculating
+   * @returns {Promise<NetWorkNextNonce>}
+   */
+  async _getNetworkNextNonce (address: string): Promise<NetWorkNextNonce> {
+    // calculate next nonce
+    // we need to make sure our base count
+    // and pending count are from the same block
+    const blockNumber: Promise<any> = await this.blockTracker.getLatestBlock()
```

Comment:
This is incorrect. `blockTracker.getLatestBlock` returns the number of the latest block as a `0x`-prefixed `string`.

So, the type of `getLatestBlock` is `() => Promise<string>`, and the type of `blockNumber` is `string`.